### PR TITLE
Slice 2.0 based on MemorySegment

### PR DESCRIPTION
**Pull request by dain**
New version of Slice based on the Java 20 preview MemorySegment APIs.

### CODE REVIEW

File: src/main/java/io/airlift/slice/Slice.java [hunk 1]
```diff
@@ -1197,30 +959,16 @@ public String toStringUtf8()
...
-            //noinspection deprecation
-            return new String(byteArray(), 0, byteArrayOffset() + index, length);
-        }
-
-        char[] chars = new char[length];
-        for (int pos = index; pos < length; pos++) {
-            chars[pos] = (char) (getByteUnchecked(pos) & 0x7F);
-        }
-        return new String(chars);
+        return toString(0, length, US_ASCII);
```
Comment (wendigo) [inline src/main/java/io/airlift/slice/Slice.java:967]: did you mean `toString(index, length, US_ASCII)` ?

**Comment by wendigo** [comment]
@dain can you share any benchmarks?

File: src/main/java/io/airlift/slice/XxHash64.java [hunk 2]
```diff
@@ -15,15 +15,21 @@
...
 import static java.lang.Long.rotateLeft;
 import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.nio.ByteOrder.LITTLE_ENDIAN;
 import static java.util.Objects.checkFromIndexSize;
-import static sun.misc.Unsafe.ARRAY_BYTE_BASE_OFFSET;
 
 public final class XxHash64
 {
+    private static final ValueLayout.OfLong LONG = ValueLayout.JAVA_LONG_UNALIGNED.withOrder(LITTLE_ENDIAN);
```
Comment (wendigo) [inline src/main/java/io/airlift/slice/XxHash64.java:29]: We have a similar declaration in `Slice`. Can we move it to Slices and reuse it between those two classes?

File: src/test/java/io/airlift/slice/TestSliceCompactFlag.java [hunk 3]
```diff
@@ -27,57 +27,50 @@
...
-        assertCompact(new Slice(booleanArray, 0, booleanArray.length));
-        assertNotCompact(new Slice(booleanArray, 0, booleanArray.length - 1));
-        assertNotCompact(new Slice(booleanArray, 1, booleanArray.length - 1));
-
         byte[] byteArray = {(byte) 0, (byte) 1, (byte) 2, (byte) 3, (byte) 4, (byte) 5};
-        assertCompact(new Slice(byteArray));
-        assertCompact(new Slice(byteArray, 0, byteArray.length));
-        assertNotCompact(new Slice(byteArray, 0, byteArray.length - 1));
-        assertNotCompact(new Slice(byteArray, 1, byteArray.length - 1));
+        assertCompact(Slices.wrappedBuffer(byteArray));
```
Comment (wendigo) [inline src/test/java/io/airlift/slice/TestSliceCompactFlag.java:31]: nit: static import wrappedBuffer and rest

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -7,7 +7,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        java: ['17', '20']
+        java: ['20']
```
Comment (wendigo) [inline .github/workflows/main.yml:10]: 21-ea

**Comment by dain** [comment]
For now, we are going to go with only supporting byte[] as a backend, so memory segment is not required.  When memory segment comes out, we can replace the last unsafe usages (copying to/from other array types) with memory segment copy.

**Pull request closed without merging**